09/tree23/node.go

```diff
@@ -0,0 +1,475 @@
+package tree23
+
+import "golang.org/x/exp/constraints"
+
+type node[T constraints.Ordered] struct {
+	parent *node[T]    // родитель узла
+	sons   [4]*node[T] // сыновья узла
+	keys   [3]T        // ключи узла
+	length int         // количество ключей
+}
+
+func (n *node[T]) Insert(x T) *node[T] {
+	if n == nil {
+		return &node[T]{
+			parent: nil,
+			keys:   [3]T{x},
+			length: 1,
+		}
+	}
+
+	if n.isLeaf() {
+		n.insertToNode(x)
+	} else if x <= n.keys[0] {
+		n.sons[0].Insert(x)
+	} else if (n.length == 1) || ((n.length == 2) && x <= n.keys[1]) {
+		n.sons[1].Insert(x)
+	} else {
+		n.sons[2].Insert(x)
+	}
+
+	return n.split()
+}
+
+func (n *node[T]) Search(x T) *node[T] {
+	if n == nil {
+		return nil
+	}
+
+	if n.findKey(x) {
+		return n
+	} else if x < n.keys[0] {
+		return n.sons[0].Search(x)
+	} else if ((n.length == 2) && (x < n.keys[1])) || (n.length == 1) {
+		return n.sons[1].Search(x)
+	} else if n.length == 2 {
+		return n.sons[2].Search(x)
+	}
+
+	return nil
+}
+
+func (n *node[T]) Remove(x T) *node[T] {
+	item := n.Search(x)
+
+	if item == nil {
+		return n
+	}
+
+	var minNode *node[T]
+	if item.keys[0] == x {
+		minNode = item.sons[1].searchMin()
+	} else {
+		minNode = item.sons[2].searchMin()
+	}
+
+	if minNode != nil {
+		var z T
+		if x == item.keys[0] {
+			z = item.keys[0]
+		} else {
+			z = item.keys[1]
+		}
+		z, minNode.keys[0] = minNode.keys[0], z
```

Comment:
второй баг тут:
ты записываешь ключ из минимальной ноды в z, но z нигде не используется, а ведь мы должны переложить этот ключ в другую ноду, иначе мы этот ключ потеряем
поэтому z тут должен быть указателем на элемент массива ключей, который мы хотим затереть, чтобы вместо него вставить ключ из минимальной ноды, тогда мы этот ключ не потеряем
и спокойно затрем именно тот, который хотим удалить
```suggestion
		var z *T
		if x == item.keys[0] {
			z = &item.keys[0]
		} else {
			z = &item.keys[1]
		}
		*z, minNode.keys[0] = minNode.keys[0], *z
```